collect_app/src/main/java/org/odk/collect/android/utilities/DownloadFormListUtils.java

```diff
@@ -61,24 +66,51 @@ public DownloadFormListUtils() {
     }
 
     public HashMap<String, FormDetails> downloadFormList(boolean alwaysCheckMediaFiles) {
-        SharedPreferences settings =
-                PreferenceManager.getDefaultSharedPreferences(
+        return downloadFormList(null, null, null, alwaysCheckMediaFiles);
+    }
+
+    public HashMap<String, FormDetails> downloadFormList(@Nullable String url, @Nullable String username,
+                                                         @Nullable String password, boolean alwaysCheckMediaFiles) {
+        SharedPreferences settings = PreferenceManager.getDefaultSharedPreferences(
                         Collect.getInstance().getBaseContext());
-        String downloadListUrl =
+
+        // Remove trailing '/'
+        while (url != null && url.endsWith("/")) {
+            url = url.substring(0, url.length() - 1);
+        }
+
+        //Todo: Make the username/password optional for cases of anonymous logins
+        String downloadListUrl = url != null ? url :
                 settings.getString(PreferenceKeys.KEY_SERVER_URL,
                         Collect.getInstance().getString(R.string.default_server_url));
         // NOTE: /formlist must not be translated! It is the well-known path on the server.
         String formListUrl = Collect.getInstance().getApplicationContext().getString(
                 R.string.default_odk_formlist);
-        String downloadPath = settings.getString(PreferenceKeys.KEY_FORMLIST_URL, formListUrl);
+
+        // When a url is supplied, we will use the default formList url
+        String downloadPath = (url != null && username != null && password != null) ? formListUrl : settings.getString(PreferenceKeys.KEY_FORMLIST_URL, formListUrl);
```

Comment:
This part should just be `url != null` now.

Reply:
On it